src/main/java/com/avispl/symphony/dal/infrastructure/management/qsc/qsysreflect/QSysReflectCommunicator.java

```diff
@@ -0,0 +1,734 @@
+			List<String> filterDeviceModelValues = handleListFilterModel();
+			if (logger.isDebugEnabled()) {
+				logger.debug(String.format("Applying device model filter with values(s): %s", filterModel));
+			}
+			List<AggregatedDevice> filteredAggregatedDevice = new ArrayList<>();
+			for (AggregatedDevice aggregatedDevice : resultAggregatedDeviceList) {
+				if (filterDeviceModelValues.contains(aggregatedDevice.getDeviceModel())) {
+					filteredAggregatedDevice.add(aggregatedDevice);
+				}
+			}
+			return filteredAggregatedDevice;
+		}
+		return resultAggregatedDeviceList;
+	}
+
+	/**
+	 * Populate normalize uptime and status message from the API
+	 */
+	private void populateDeviceUptimeAndStatusMessage(List<AggregatedDevice> resultAggregatedDeviceList) {
+		for (int i = 0; i < resultAggregatedDeviceList.size(); i++) {
+			AggregatedDevice aggregatedDevice = resultAggregatedDeviceList.get(i);
+			Map<String, String> properties = aggregatedDevice.getProperties();
+			properties.put(QSysReflectConstant.DEVICE_UPTIME, handleNormalizeUptime(Long.parseLong(properties.get(QSysReflectConstant.DEVICE_UPTIME))));
+			properties.put(QSysReflectConstant.DEVICE_STATUS_MESSAGE, deviceStatusMessageMap.get(aggregatedDevice.getDeviceId()));
+			resultAggregatedDeviceList.set(i, aggregatedDevice);
+		}
+	}
+
+	/**
+	 * Retrieve aggregated devices and system information data -
+	 * and set next device/system collection iteration timestamp
+	 */
+	private void retrieveInfo() {
+		long currentTimestamp = System.currentTimeMillis();
+		if (validDeviceMetaDataRetrievalPeriodTimestamp > currentTimestamp) {
+			if (logger.isDebugEnabled()) {
+				logger.debug(String.format("Aggregated devices data and system information retrieval is in cool down. %s seconds left",
+						(validDeviceMetaDataRetrievalPeriodTimestamp - currentTimestamp) / 1000));
+				if (!aggregatedDeviceList.isEmpty()) {
+					logger.debug(String.format("Old fetched devices list: %s", aggregatedDeviceList));
+				}
+				if (!systemResponseList.isEmpty()) {
+					logger.debug(String.format("Old system information list: %s", systemResponseList));
+				}
+			}
+			return;
+		}
+		validDeviceMetaDataRetrievalPeriodTimestamp = currentTimestamp + deviceMetaDataRetrievalTimeout;
+		retrieveDevices();
+		if (logger.isDebugEnabled()) {
+			logger.debug(String.format("New fetched devices list: %s", aggregatedDeviceList));
+		}
+		retrieveSystemInfo();
+		if (logger.isDebugEnabled()) {
+			logger.debug(String.format("New fetched system information list: %s", systemResponseList));
+		}
+		nextDevicesCollectionIterationTimestamp = System.currentTimeMillis();
+	}
+
+	/**
+	 * Get list of device every 30 seconds
+	 * API Endpoint: /cores
+	 * Success: Return a list of devices(cores) within the organization
+	 */
+	private void retrieveDevices() {
+		try {
+			Map<String, PropertiesMapping> mapping = new PropertiesMappingParser().loadYML("qsysreflect/model-mapping.yml", getClass());
```

Comment:
It can change if the end-user enters Q-Sys platform and adds the device. That's why I call it every interval.

Reply:
Sure, but the mapping itself (in yml file) won't change, would it? The aggregatedDeviceProcessor will keep the same mapping information there, which target API doesn't have control over. So the json on the Q-Sys would change, but the mapping and the state of the processor will stay the same, so there's no need to re-initialize it every time (what json will change is what happens on line 577 - aggregatedDeviceProcessor.extractDevices(devices), but the aggregatedDeviceProcessor state is pretty much constant per adapter instance since we don't change yml on fly)

Reply:
Updated. I will initialize aggregatedDeviceProcessor in the constructor.